base/sysinfo.jl

```diff
@@ -321,4 +323,83 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys.isexecutable(path::AbstractString)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function isexecutable(path::AbstractString)
+    @static if iswindows()
+        return true
+    else
+        # We use `access()` and `X_OK` to determine if a given path is
+        # executable by the current user.  `X_OK` comes from `unistd.h`.
+        X_OK = 0x01
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+
+"""
+    Sys.which(program_name::AbstractString)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns the absolute
+path. Raises `ArgumentError` if no such program is available.  If a relative or
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::AbstractString)
+    # If prog has a slash, we know the user wants to determine whether the given
+    # file exists and is executable, and to not search the `PATH`.  Note that
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static iswindows() ? ('/', '\\') : ('/', )
+
+    if any(occursin.(dirseps, program_name))
+        # If this file does not even exist, fail out
+        if !isfile(program_name)
+            throw(ArgumentError("$program_name does not exist"))
+        end
+
+        # If it does exist, check that it's executable or fail out
+        if !isexecutable(program_name)
+            throw(ArgumentError("$program_name is not executable"))
+        end
+
+        # If it all checks out, return the abspath
+        return abspath(program_name)
+    end
+
+    # If we have been given just a program name (not a relative or absolute
+    # path) then we should search `PATH` for it here:
+    path = get(ENV, "PATH", "")
+    pathsep = @static iswindows() ? ';' : ':'
+
+    # On windows, if the program name does not already have an extension,
+    # we need to tack all the values in PATHEXT onto the end and check for
```

Comment:
Nope, see how libuv implements this function. IIRC, the distinction is that PATHEXT is for DOS, not the Shell32.

Reply:
Which function are you referring to? What should I look for in `libuv`?

I'm not sure what you mean in stating that `PATHEXT` is only for DOS; when I try to `run()` a command, I expect it to work like I've typed it into `cmd.exe`, and `cmd.exe` does pay attention to `PATHEXT`.

Reply:
See my libuv link in the other comment thread

Reply:
You want to match the semantics of `search_path` in libuv's `src/win/process.c`, I guess, since this is what `run` uses.

Unfortunately, the `search_path` function is not exported.  But since we use our own libuv fork anyway, maybe we should just modify it to export a `uv_win_search_path` function on Windows?

Reply:
Libuv says:
> Furthermore, it tries to follow the semantics that cmd.exe, with this exception that PATHEXT environment variable isn't used. Since CreateProcess can start only .com and .exe files, only those extensions are tried. This  behavior equals that of msvcrt's spawn functions.

Reply:
Note also that libuv writes *CMD does not trim leading/trailing whitespace from path/pathex entries*